### PR TITLE
ARROW-13989: [C++] Add support for month-day-nano interval to compute functions

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc [hunk 1]
```diff
@@ -1302,7 +1304,8 @@ struct CopyFixedWidth<BooleanType> {
 };
 
 template <typename Type>
-struct CopyFixedWidth<Type, enable_if_number<Type>> {
+struct CopyFixedWidth<
+    Type, enable_if_t<is_number_type<Type>::value || is_interval_type<Type>::value>> {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_if_else.cc:1308]: Out of curiosity, why `is_interval_type` here while IfElseFunctor only allows month-day-nano intervals?

File: cpp/src/arrow/compute/kernels/vector_replace_test.cc [hunk 2]
```diff
@@ -580,6 +588,80 @@ TEST_F(TestReplaceDayTimeInterval, ReplaceWithMask) {
...
+  this->Assert(ReplaceWithMask, this->array("[]"), this->mask("[]"),
+               this->scalar("[7, 0, 8]"), this->array("[]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [3, 4, -2]]"),
+               this->mask("[true, true]"), this->scalar("[7, 0, 8]"),
+               this->array("[[7, 0, 8], [7, 0, 8]]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [3, 4, -2]]"),
+               this->mask("[true, true]"), this->scalar("null"),
+               this->array("[null, null]"));
+  this->Assert(ReplaceWithMask, this->array("[[1, 2, 4], [3, 4, -2], [-5, 6, 7]]"),
+               this->mask("[true, false, null]"), this->scalar("[7, 0, 8]"),
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_replace_test.cc:661]: Slightly unrelated to this PR, but it seems that amongst the `ReplaceWithMask` tests, the mask is always of the form (regex) `true* (false|null)*`, meaning that the replacement index is always trivially equal to the source index. Is that deliberate?

**Review by pitrou — APPROVED** [review]
+1, just two questions

File: cpp/src/arrow/compute/kernels/scalar_if_else.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/scalar_if_else.cc:1308]: This helper is used by coalesce/choose/case_when so this is part of adding support for all of those functions.

if_else should support all the interval types. It uses GenerateTypeAgnosticPrimitive, so DayTime and Month intervals are handled as integers.

File: cpp/src/arrow/compute/kernels/vector_replace_test.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/vector_replace_test.cc:661]: It's not deliberate. I'll update the tests to avoid this.

**Pull request closed without merging**